src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -362,5 +413,113 @@ private void Cancel()
+        }
+
+        private ValueTask<ReadResult> CompleteReadAsync(ValueTask<int> task, CancellationToken cancellationToken, CancellationTokenRegistration reg)
+        {
+            Debug.Assert(_readInProgress, "Read not in progress");
+
+            _readCancellation = cancellationToken;
+            _readRegistration = reg;
+
+            _readAwaiter = task.GetAwaiter();
+
+            return new ValueTask<ReadResult>(this, _readMrvts.Version);
+        }
+
+        private void OnReadCompleted()
+        {
+            try
+            {
+                int length = _readAwaiter.GetResult();
+
+                Debug.Assert(length + _readTail.End <= _readTail.AvailableMemory.Length);
+
+                _readTail.End += length;
+                _bufferedBytes += length;
+
+                if (length == 0)
+                {
+                    _isStreamCompleted = true;
+                }
+
+                _readMrvts.SetResult(new ReadResult(GetCurrentReadOnlySequence(), isCanceled: false, _isStreamCompleted));
+            }
+            catch (OperationCanceledException oce)
+            {
+                // Get the source before clearing (and replacing)
+                CancellationTokenSource tokenSource = InternalTokenSource;
+                ClearCancellationToken();
+                if (tokenSource.IsCancellationRequested && !_readCancellation.IsCancellationRequested)
+                {
+                    // Catch cancellation and translate it into setting isCanceled = true
+                    _readMrvts.SetResult(new ReadResult(GetCurrentReadOnlySequence(), isCanceled: true, _isStreamCompleted));
+                }
+                else
+                {
+                    _readMrvts.SetException(oce);
+                }
+            }
+            catch (Exception ex)
+            {
+                _readMrvts.SetException(ex);
+            }
+            finally
+            {
+                _readRegistration.Dispose();
+                _readRegistration = default;
+            }
+        }
+
+        ReadResult IValueTaskSource<ReadResult>.GetResult(short token)
+        {
+            ValidateReading();
+            ReadResult result = _readMrvts.GetResult(token);
+
+            _readCancellation = default;
+            _readAwaiter = default;
+            _readMrvts.Reset();
+
+            Debug.Assert(_readInProgress);
+            _readInProgress = false;
+
+            return result;
+        }
+
+        ValueTaskSourceStatus IValueTaskSource<ReadResult>.GetStatus(short token)
+            => _readMrvts.GetStatus(token);
+
+        void IValueTaskSource<ReadResult>.OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            ValidateReading();
+            _readMrvts.OnCompleted(continuation, state, token, flags);
+
+            if ((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) != 0)
```

Comment:
This should ignore the flags and always use `UnsafeOnCompleted`, the `ManualResetValueTaskSource` already reads the flags and will save/restore the appropriate context when SetResult is called.

Reply:
Changed; though I need to resolve the conflics

Reply:
Done